### PR TITLE
Add optimized distiance functions for PowerPC

### PR DESCRIPTION
**Pull request by carll99**
Files src/simd/distances_powerpc.cc, src/simd/distances_powerpc.h are added with powerpc optimized versions of the various functions.

File src/simd/hook.cc is updated to point to the new PowerPC versions of the functions.

### CODE REVIEW

**Comment by carll99** [comment]
Based on the comments I got from generating the PR, it sounds like I am supposed to set the kind label?  Not sure how that should have been done.  Secondly, this adds specific PowerPC support which I believe makes it  kind/feature?

File: src/simd/distances_powerpc.cc [hunk 1]
```diff
@@ -0,0 +1,540 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (alexanderguzhva) [inline src/simd/distances_powerpc.cc:2]: Facebook, not Zilliz in a license header?

File: src/simd/distances_powerpc.cc [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+#include <cmath>
+
+#define FLOAT_VEC_SIZE 4
+#define INT32_VEC_SIZE 4
+#define INT8_VEC_SIZE  16
+
+namespace faiss {
+
+float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d) {
```
Comment (alexanderguzhva) [inline src/simd/distances_powerpc.cc:23]: please call the function `fvec_L2sqr_ppc`, not `fvec_L2sqr_ref_ppc`. 
Same for all other functions.

File: src/simd/distances_powerpc.cc [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+    float res = 0;
+    size_t base;
+
+    vector float *vx, *vy;
+    vector float vtmp = {0, 0, 0, 0};
+    vector float vres = {0, 0, 0, 0};
+
+    base = (d / FLOAT_VEC_SIZE) * FLOAT_VEC_SIZE;
+
+    for (size_t i = 0; i < base; i = i + FLOAT_VEC_SIZE) {
```
Comment (alexanderguzhva) [inline src/simd/distances_powerpc.cc:37]: `for (size_t i = 0; i < base; i += FLOAT_VEC_SIZE) {`
same comment applies to loops in other functions

File: src/simd/distances_powerpc.cc [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+#define INT8_VEC_SIZE  16
+
+namespace faiss {
+
+float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d) {
+    /* Vector implmentaion uses vector size of FLOAT_VEC_SIZE.  If the input
+       array size is not a power of FLOAT_VEC_SIZE, do the remaining elements
+       in scalar mode.  */
+    size_t i;
```
Comment (alexanderguzhva) [inline src/simd/distances_powerpc.cc:27]: please get rid of this local variable.
same comment applies to other functions

File: src/simd/distances_powerpc.cc [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    for (size_t i = 0; i < base; i = i + FLOAT_VEC_SIZE) {
+        vx = (vector float *)(&x[i]);
+        vy = (vector float *)(&y[i]);
+
+        vtmp = vx[0] - vy[0];
+        vres += vtmp * vtmp;
+    }
+
+    /* Handle any remaining data elements */
+    for (i = base; i < d; i++) {
```
Comment (alexanderguzhva) [inline src/simd/distances_powerpc.cc:46]: `for (size_t i = base; i < d; i++) {`
same comment applies to loops in other functions

File: src/simd/distances_powerpc.cc [hunk 6]
```diff
@@ -0,0 +1,540 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#if defined(__powerpc__)
```
Comment (alexanderguzhva) [inline src/simd/distances_powerpc.cc:8]: I just want to confirm that all knowhere unit tests passed, correct?

**Comment by carll99** [comment]
Alexander:

Thank you for reviewing the code.  I will make the updates and we will retest the patch.  Per your comment at the bottom, yes the changes were tested on a PowerPC processor to verify the code compiles, executes and produces the same results as the reference functions.

Just want to verify the procedure for posting the updated version.  I am assuming that once I have resynched my github with the latest upstream code and my updated patch that I would create a new pull request.  Not sure if there is anything I need to do to cancel the initial pull request.

Thanks for the help.

                                       Carl Love
________________________________
From: Alexander Guzhva ***@***.***>
Sent: Wednesday, October 16, 2024 4:23 AM
To: zilliztech/knowhere ***@***.***>
Cc: Carl Love ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [zilliztech/knowhere] Add optimized distiance functions for PowerPC (PR #894)

@ alexanderguzhva commented on this pull request. In src/simd/distances_powerpc. cc: > @@ -0,0 +1,540 @@ +/** + * Copyright (c) Facebook, Inc. and its affiliates. Facebook, not Zilliz in a license header? In src/simd/distances_powerpc. cc: >


@alexanderguzhva commented on this pull request.

________________________________

In src/simd/distances_powerpc.cc<https://github.com/zilliztech/knowhere/pull/894#discussion_r1802897010>:

> @@ -0,0 +1,540 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.


Facebook, not Zilliz in a license header?

________________________________

In src/simd/distances_powerpc.cc<https://github.com/zilliztech/knowhere/pull/894#discussion_r1802899515>:

> +#if defined(__powerpc__)
+
+#include <altivec.h>   /* Required for the Power GCC built-ins  */
+
+#include "distances_powerpc.h"
+
+#include <cmath>
+
+#define FLOAT_VEC_SIZE 4
+#define INT32_VEC_SIZE 4
+#define INT8_VEC_SIZE  16
+
+namespace faiss {
+
+float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d) {


please call the function fvec_L2sqr_ppc, not fvec_L2sqr_ref_ppc.
Same for all other functions.

________________________________

In src/simd/distances_powerpc.cc<https://github.com/zilliztech/knowhere/pull/894#discussion_r1802900475>:

> +float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d) {
+    /* Vector implmentaion uses vector size of FLOAT_VEC_SIZE.  If the input
+       array size is not a power of FLOAT_VEC_SIZE, do the remaining elements
+       in scalar mode.  */
+    size_t i;
+    float res = 0;
+    size_t base;
+
+    vector float *vx, *vy;
+    vector float vtmp = {0, 0, 0, 0};
+    vector float vres = {0, 0, 0, 0};
+
+    base = (d / FLOAT_VEC_SIZE) * FLOAT_VEC_SIZE;
+
+    for (size_t i = 0; i < base; i = i + FLOAT_VEC_SIZE) {


for (size_t i = 0; i < base; i += FLOAT_VEC_SIZE) {
same comment applies to loops in other functions

________________________________

In src/simd/distances_powerpc.cc<https://github.com/zilliztech/knowhere/pull/894#discussion_r1802900629>:

> +#include "distances_powerpc.h"
+
+#include <cmath>
+
+#define FLOAT_VEC_SIZE 4
+#define INT32_VEC_SIZE 4
+#define INT8_VEC_SIZE  16
+
+namespace faiss {
+
+float
+fvec_L2sqr_ref_ppc(const float* x, const float* y, size_t d) {
+    /* Vector implmentaion uses vector size of FLOAT_VEC_SIZE.  If the input
+       array size is not a power of FLOAT_VEC_SIZE, do the remaining elements
+       in scalar mode.  */
+    size_t i;


please get rid of this local variable.
same comment applies to other functions

________________________________

In src/simd/distances_powerpc.cc<https://github.com/zilliztech/knowhere/pull/894#discussion_r1802907583>:

> +    vector float *vx, *vy;
+    vector float vtmp = {0, 0, 0, 0};
+    vector float vres = {0, 0, 0, 0};
+
+    base = (d / FLOAT_VEC_SIZE) * FLOAT_VEC_SIZE;
+
+    for (size_t i = 0; i < base; i = i + FLOAT_VEC_SIZE) {
+        vx = (vector float *)(&x[i]);
+        vy = (vector float *)(&y[i]);
+
+        vtmp = vx[0] - vy[0];
+        vres += vtmp * vtmp;
+    }
+
+    /* Handle any remaining data elements */
+    for (i = base; i < d; i++) {


for (size_t i = base; i < d; i++) {
same comment applies to loops in other functions

________________________________

In src/simd/distances_powerpc.cc<https://github.com/zilliztech/knowhere/pull/894#discussion_r1802908959>:

> @@ -0,0 +1,540 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#if defined(__powerpc__)


I just want to confirm that all knowhere unit tests passed, correct?

**Comment by alexanderguzhva** [comment]
@carll99 just force push the new changes to override this PR with a better version

**Comment by liliu-z** [comment]
Hi @carll99 
Due to the limited resource, we actually doesn't officially support PowerPC, which means the official release and new features will not take care of the PowerPC specific scenario. 
However, we are welcome for the contribution and maintain from the community. Do you mind to share with us more about the use case to help us know more about the importance of PowerPC?

**Comment by alexanderguzhva** [comment]
@carll99 please accept DCO and run clang-format and I'll lgtm the PR. Thanks.

**Comment by carll99** [comment]
I updated the patch and the team is working on retesting it.  

I forwarded the comment from liliu-z to the PowerPC and we will work on the response.  

The DCO, looks like the description for DCO says the project wants a sign-off line.  I will include that with the next version.  

I will check how to run the clang-format  tool.  I did a quick google to see what it is.   

Thanks for the feedback.

**Comment by alexanderguzhva** [comment]
@carll99 Please carefully read the error message https://github.com/zilliztech/knowhere/pull/894/checks?check_run_id=32817538304, all instructions are there
Also, would you please update a PR to include your changes?
Thanks

**Review by foxspy — APPROVED** [review]

**Pull request merged**